### ISSUE TITLE
Allow excluding entire modules/packages

### ISSUE THREAD

**Issue by nedbat** [coveragepy/coveragepy#118]
*Originally reported by* **Ram Rachum (Bitbucket: [coolrr](https://bitbucket.org/coolrr), GitHub: [coolrr](https://github.com/coolrr))**

----------------------------------------

Allow excluding entire modules/packages from coverage measurement by including some identifiable string in their source, similarly to "pragma: no cover".

Here's a suggestion on how to do this, feel free to suggest alternatives: A line "pragma: no module cover" in a module will cause it not to be measured. A "pragma: no package cover" in an `__init__.py` file will cause the entire package, (including any submodules, subpackages, etc.) not to be measured.

----------------------------------------
- Bitbucket: https://bitbucket.org/ned/coveragepy/issue/118

**Comment by nedbat** [comment]
Issue #300 was marked as a duplicate of this issue.

**Comment by nedbat** [comment]
*Original comment by* **René Fleschenberg (Bitbucket: [rfleschenberg](https://bitbucket.org/rfleschenberg), GitHub: [rfleschenberg](https://github.com/rfleschenberg))**

----------------------------------------

Is there a reason why we cannot just use `pragma: no cover` at module level to exclude the whole module from coverage?

**Comment by nedbat** [comment]
I haven't looked into what it would take to implement this.  Ram's original proposal included a new pragma for `__init__.py` to exclude the whole package.  That would need to be different than the pragma for a single module, I think.

**Comment by nedbat** [comment]
As of commit 1fe897d740ebc67bd1f6493bcc889846f16abf10, you can do this with a multi-line exclusion regex:
```ini
[report]
exclude_also =
    (?s)\A.*# pragma: exclude file.*\Z
```
and then put the comment `# pragma: exclude file` in the file somewhere.

**Comment by nedbat** [comment]
This is now released as part of [coverage 7.6.0](https://pypi.org/project/coverage/7.6.0).
